### ISSUE TITLE
meson wrongly assumes that x86_64-linux-gnu can run x86_64-linux-gnux32

### ISSUE THREAD

**Issue by helmutg** [mesonbuild/meson#9845]
**Describe the bug**
When performing a cross build with a crossfile where the build architecture is x86_64-linux-gnu and the host architecture is x86_64-linux-gnux32 `meson.has_exe_wrapper()` returns `true`. However, many distribution kernels (including Debian) disable the x32 ABI by default to reduce the attack surface. Actually running executables fails with an `Exec format error`.

**To Reproduce**
meson.build:

    project('foo')
    message(meson.has_exe_wrapper())

crossfile (from `debcrossgen` by @jpakkane):

    [binaries]
    c = '/usr/bin/x86_64-linux-gnux32-gcc'
    cpp = '/usr/bin/x86_64-linux-gnux32-g++'
    ar = '/usr/bin/x86_64-linux-gnux32-ar'
    strip = '/usr/bin/x86_64-linux-gnux32-strip'
    objcopy = '/usr/bin/x86_64-linux-gnux32-objcopy'
    ld= '/usr/bin/x86_64-linux-gnux32-ld'
    [properties]
    [host_machine]
    system = 'linux'
    cpu_family = 'x86_64'
    cpu = 'x32'
    endian = 'little'

It will output:

    Message: true

**Expected behavior**

For the presented combination, `meson.has_exe_wrapper()` should return `false`. The issue is wider though. Some `armv8` systems are no longer capable of running `armv7` code. That's essentially the same issue. More generally, the heuristic in `machine_info_can_run` is too simple and has both false positives (such as this case) and false negatives (for instance when qemu has been set up with binfmt support). It would be good to simply run a host executable and see whether it works instead.

**system parameters**
* what operating system: Debian unstable
* Python 3.9.10
* meson 0.61.1-1

**discussion**
The issue has likely been caused by #516. Thanks to @eli-schwartz and @dcbaker for figuring this out.

**Comment by eli-schwartz** [comment]
I agree this heuristic should be fixed, there doesn't seem to be any particularly good reason to do it the way it is right now (maybe it was just not thought to be necessary, but it is too unreliable).

The general workaround is, of course, to determine outside of meson whether those executables will run, and then override that information in the meson cross file using the `needs_exe_wrapper` property.

**Comment by helmutg** [comment]
The `debcrossgen` utility that is generally used to generate the crossfile is in the same position here. It would have to add a heuristic and we'd just be moving the problem from meson upstream to the Debian packaging.

**Comment by eli-schwartz** [comment]
Yep, indeed, workarounds are awkward (hence the name). Just wanted to make note of it for others stumbling on the issue. :)

**Comment by nirbheek** [comment]
I like the idea of compiling and running a host machine executable to check whether it can be run natively without a wrapper.
